pkg/apiserver/metrics/deployments.go

```diff
@@ -0,0 +1,168 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"k8s.io/kubernetes/pkg/api/rest"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/net/context"
+)
+
+var (
+	variableLabels = []string{"name", "namespace"}
+
+	deploymentStatusObservedGenerationDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_observed_generation",
+		"The observed generation of this deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_replicas",
+		"The number of replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUpdatedReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_updated_replicas",
+		"The number of updated replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusAvailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_available_replicas",
+		"The number of available replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUnavailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_unavailable_replicas",
+		"The number of unavailable replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+)
+
+// DeploymentCollector collects various metrics of deployment objects.
+type DeploymentCollector struct {
+	Storage rest.StandardStorage
+}
+
+// Returns a new instance of DeploymentCollector.
+func NewDeploymentsCollector(s rest.StandardStorage) *DeploymentCollector {
+	return &DeploymentCollector{
+		Storage: s,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- deploymentStatusObservedGenerationDesc
+	ch <- deploymentStatusReplicasDesc
+	ch <- deploymentStatusUpdatedReplicasDesc
+	ch <- deploymentStatusAvailableReplicasDesc
+	ch <- deploymentStatusUnavailableReplicasDesc
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Collect(ch chan<- prometheus.Metric) {
+	var err error
+	deployments, err := c.Storage.List(context.Background(), nil)
+	if err != nil {
+		glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+	}
+
+	for _, d := range deployments.(*extensions.DeploymentList).Items {
+		err = addMetrics(ch, d)
+		if err != nil {
+			glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+		}
+	}
+}
+
+// addMetrics adds Metrics for a single deployment to the specified channel
+func addMetrics(ch chan<- prometheus.Metric, d extensions.Deployment) error {
```

Comment:
Generic resource level metrics are not unreasonable to add to core Kubernetes, especially if they are identical across all resources.  Domain level metrics I think do tend to belong with the controllers, which have access to this info as well.

Regarding size of cluster, we have multiple production clusters of Kube exceeding 5k deployments today (as well as similar numbers of other resources) that are very dense, but not very wide in terms of nodes.  So things we might not have worried about last year w.r.t. metric density are more relevant now.

Also, the security aspect of metrics and monitoring at least needs to be addressed - if metrics exposes name of resource, that falls under the security policy rules and so sig-auth needs to at least have a chance to review a proposal.  We have mechanisms to control access to apiserver endpoints, but controller endpoints are not protected the same way today (they probably should).

Finally, DoS protection and limiting is mostly handled by quota and limit range which provide object count and resource consumption controls.  However, they do not (today) provide rate control on changes / per second.  So the quota usage object should already have some of the count metrics for "who is doing what", if you are concerned about enforcement.  If you are only concerned about data, then metrics is probably sufficient.